posts/2017-07-24-anatomy-of-a-serverless-app.md

```diff
@@ -0,0 +1,669 @@
+
+```
+$ curl -X POST https://yt1i5ydiu4.execute-api.us-east-1.amazonaws.com/prod/email -d '{"to_address":"your@email.com"}'
+
+HTTP/1.1 202 Accepted
+...
+...
+
+{"message":"Request to send email is successful.","input":{"id":"<20170721005711.6305.4865866169A6F957@sandbox77a6b258412d4bb9a2d12abeb33ac01e.mailgun.org>","message":"Queued. Thank you."}}
+```
+Again, another email is successfully sent.
+
+### Error Handling
+
+To have our code gracefully handle errors and bad input, we added some validation checks. The code makes sure that if bad input data is passed in, the function returns an error even before calling the `mailgun` API method. That not only saves time & increases response time but potentailly saves money at the Mailgun side. Let's try out some error edge cases and see the results:
+
+**No Data**
+
+```
+$ curl -i -X POST https://yt1i5ydiu4.execute-api.us-east-1.amazonaws.com/prod/email
+
+{"message":"Bad input data or missing email address.","input":null}
+```
+
+**Empty JSON**
+
+```
+$ curl -i -X POST -d '{}' https://yt1i5ydiu4.execute-api.us-east-1.amazonaws.com/prod/email
+
+{"message":"Bad input data or missing email address.","input":"{}"}
+```
+
+**Bad Input Structure**
+
+```
+$ curl -i -X POST -d '{"junk"}' https://yt1i5ydiu4.execute-api.us-east-1.amazonaws.com/prod/email
+
+{"message":"Bad input data or missing email address.","input":"{\"junk\"}"}
+```
+
+**No Email Address**
+
+```
+$ curl -i -X POST -d '{"to_address":""}' https://yt1i5ydiu4.execute-api.us-east-1.amazonaws.com/prod/email
+
+{"message":"Bad input data or missing email address.","input":"{\"to_address\":\"\"}"}
+```
+
+**Malformed Email Address**
+
+```
+$ curl -i -X POST -d '{"to_address":"junk"}' https://yt1i5ydiu4.execute-api.us-east-1.amazonaws.com/prod/email
+
+{"message": "Internal server error"}
+```
+`curl` kind of barfs, as the function causes an exception in this case. We can see that if we call the function with the framework:
+
+```
+$ sls invoke --function send --data '{"body":"{\"to_address\":\"junk\"}"}' --log
+
+{
+    "errorMessage": "'to' parameter is not a valid address. please check documentation",
+    "errorType": "Error",
+    "stackTrace": [
+        ...
+    ]
+}
+--------------------------------------------------------------------
+START RequestId: 0eb38aeb-6d9d-11e7-9787-2996c7e3fdb0 Version: $LATEST
+2017-07-20 22:44:58.544 (+00:00)	0eb38aeb-6d9d-11e7-9787-2996c7e3fdb0	{ Error: 'to' parameter is not a valid address. please check documentation
+...
+...
+2017-07-20 22:44:58.545 (+00:00)	0eb38aeb-6d9d-11e7-9787-2996c7e3fdb0	{"errorMessage":"'to' parameter is not a valid address. please check documentation","errorType":"Error","stackTrace":
+...
+...
+END RequestId: 0eb38aeb-6d9d-11e7-9787-2996c7e3fdb0
+REPORT RequestId: 0eb38aeb-6d9d-11e7-9787-2996c7e3fdb0	Duration: 339.84 ms	Billed Duration: 400 ms 	Memory Size: 1024 MB	Max Memory Used: 41 MB
+
```

Comment:
It's nice to show how unhandled errors manifest themselves w/ Lambda, but don't know that we want to encourage this. Maybe add a section to refactor and catch that error and return something meaningful to the user?